contracts/StakingVault/base/RibbonVault.sol

```diff
@@ -0,0 +1,575 @@
+
+  /************************************************
+   *  IMMUTABLES & CONSTANTS
+   ***********************************************/
+
+  /// @notice RBN 0x6123b0049f904d730db3c36a31167d9d4121fa6b
+  address public immutable RBN;
+  /// @notice sRBN
+  ISRBN public immutable sRBN;
+
+  /************************************************
+   *  EVENTS
+   ***********************************************/
+
+  event Deposit(address indexed account, uint256 amount, uint256 round);
+
+  event InitiateWithdraw(
+    address indexed account,
+    uint256 shares,
+    uint256 round
+  );
+
+  event Redeem(address indexed account, uint256 share, uint256 round);
+
+  event CapSet(uint256 oldCap, uint256 newCap, address manager);
+
+  event Withdraw(address indexed account, uint256 amount, uint256 shares);
+
+  /************************************************
+   *  CONSTRUCTOR & INITIALIZATION
+   ***********************************************/
+
+  /**
+   * @notice Initializes the contract with immutable variables
+   * @param _rbn is the RBN contract
+   * @param _srbn is the sRBN contract
+   */
+  constructor(address _rbn, address _srbn) {
+    require(_rbn != address(0), "!_rbn");
+    require(_srbn != address(0), "!_srbn");
+
+    RBN = _rbn;
+    sRBN = ISRBN(_srbn);
+  }
+
+  /**
+   * @notice Initializes the Vault contract with storage variables.
+   */
+  function baseInitialize(
+    address _owner,
+    address _keeper,
+    address[] memory _vaultAssets,
+    Vault.VaultParams calldata _vaultParams
+  ) internal initializer {
+    VaultLifecycle.verifyInitializerParams(
+      _owner,
+      _keeper,
+      _vaultAssets,
+      _vaultParams
+    );
+
+    __ReentrancyGuard_init();
+    __Ownable_init();
+    transferOwnership(_owner);
+
+    keeper = _keeper;
+
+    vaultParams = _vaultParams;
+
+    vaultState.round = 1;
+
+    for (uint256 i = 0; i < _vaultAssets.length; i++) {
+      vaultAssets.push(_vaultAssets[i]);
+    }
+  }
+
+  /**
+   * @dev Throws if called by any account other than the keeper.
+   */
+  modifier onlyKeeper() {
+    require(msg.sender == keeper, "!keeper");
+    _;
+  }
+
+  /************************************************
+   *  SETTERS
+   ***********************************************/
+
+  /**
+   * @notice Sets the new keeper
+   * @param newKeeper is the address of the new keeper
+   */
+  function setNewKeeper(address newKeeper) external onlyOwner {
+    require(newKeeper != address(0), "!newKeeper");
+    keeper = newKeeper;
+  }
+
+  /**
+   * @notice Adds new asset vaults fees
+   * @param newVaultAssets is the list of new addresses
+   */
+  function addNewVaultAssets(address[] memory newVaultAssets)
+    external
+    onlyOwner
+  {
+    require(
+      newVaultAssets.length > 0 && newVaultAssets[0] != address(0),
+      "!newVaultAssets"
+    );
+
+    for (uint256 i = 0; i < newVaultAssets.length; i++) {
+      vaultAssets.push(newVaultAssets[i]);
+    }
```

Comment:
This doesn't check if the `newVaultAssets` already exists in the `vaultAssets` array. Two ways to proceed with this:

1. Do a for-loop search

```solidity
for (vaultAsset in vaultAssets) {
  for (newVaultAsset in newVaultAssets) {
    require(vaultAsset != newVaultAsset)
  }
}
```

2. Change the data structure to be mapping instead

```solidity
mapping(address => bool) public vaultAssets;

for (uint256 i = 0; i < newVaultAssets.length; i++) {
  require(!vaultAssets[newVaultAssets[i]], "Vault asset already exists");
  vaultAssets[newVaultAssets[i]] = true;
}
```

I prefer the second method more :-)

Reply:
the second method will run into issues in market module. how will we know which assets to sell for rbn?

I recommend adding a second data structure `mapping(address => bool) public vaultAssets;`